src/rdkafka_admin.c

```diff
@@ -3257,13 +3266,40 @@ rd_kafka_DeleteRecords_response_merge (rd_kafka_op_t *rko_fanout,
         rd_assert(rko_partial->rko_evtype ==
                   RD_KAFKA_EVENT_DELETERECORDS_RESULT);
 
-        /* Partitions from the DeleteRecordsResponse */
-        partitions = rd_list_elem(&rko_partial->rko_u.admin_result.results, 0);
-
-        /* Partitions (offsets) from the DeleteRecords() call */
+        /* All partitions (offsets) from the DeleteRecords() call */
         respartitions = rd_list_elem(&rko_fanout->rko_u.admin_request.
                                      fanout.results, 0);
 
+        if (rko_partial->rko_err) {
+                /* If there was a request-level error, set the error on
+                 * all requested partitions for this request. */
+                const rd_kafka_topic_partition_list_t *reqpartitions;
+                rd_kafka_topic_partition_t *reqpartition;
+
+                /* Partitions (offsets) from this DeleteRecordsRequest */
+                reqpartitions = rd_list_elem(&rko_partial->rko_u.
+                                             admin_result.args, 0);
+
+                RD_KAFKA_TPLIST_FOREACH(reqpartition, reqpartitions) {
+                        rd_kafka_topic_partition_t *respart;
+
+                        /* Find result partition */
+                        respart = rd_kafka_topic_partition_list_find(
+                                respartitions,
+                                reqpartition->topic,
+                                reqpartition->partition);
+
+                        rd_assert(respart || !*"respart not found");
```

Comment:
if i'm reading this correctly, this would occur as a result of the broker returning an unexpected set of partitions. I don't think that should be an assert, it should be an error.

Reply:
There are alot of partition lists here, so the confusion is real.

respartitions = all partitions in the DeleteRecords() call. These are the result partitions that will be provided back to the app with per-partition errors.
reqpartitions = the partitions in this particular DeleteRecordsRequest. A (copied) subset of respartitions.
partitions = the partitions returned by the broker in the DeleteRecordsResponse.

If there's a mismatch between respartitions and reqpartitions there's a bug in the client, thus the assert.
If there's a mismatch between respartitions and partitions there's a bug in the broker, not an assert.